### ISSUE TITLE
xDNSServerForwarder errors on test if DNS is not installed. Add error handling.

### ISSUE THREAD

**Issue by GoateePFE** [PowerShell/xDnsServer#45]
xDNSServerForwarder (as well as other resource in the module) need error handling logic to gracefully return FALSE on the Test function when DNS is not installed. This is particularly needed when DSC is used to audit a server config without applying.

**Comment by Persistent13** [comment]
@GoateePFE I'd be willing to add this functionality.
What do you believe the best way to do this would be?

I'm not sure how to best handle the error.
I'm thinking to update the get function to return an empty IPAddresses array, test should then not throw an error.

**Comment by iainbrighton** [comment]
I'm sure there's an Assert-Module function in the xActiveDirectory module that you can copy/duplicate!

**Comment by johlju** [comment]
Label this as help wanted to that someone in the community can run with this.

**Comment by 1800Zeta** [comment]
https://github.com/PowerShell/xDnsServer/blob/dev/DSCResources/Helper.psm1

This Helper Module includes
# Internal function to assert if the role specific module is installed or not
function Assert-Module

So just need to ensure that the others include and use it

**Comment by rchristman89** [comment]
Assert-Module will still error out on Get if the module isn't installed. I think Assert-Module still needs to be added to all of the resources in xDnsServer. 

For not erroring out on Get we could do a try catch:
```powershell
try
    {
        Assert-Module -ModuleName 'DNS'
    }
catch
{
    Write-Warning -Message 'DNS module is not installed and this resource will not work properly.'
    return @{
         Ensure = 'Absent'
         Prop    = $null
    }
}
```

**Comment by johlju** [comment]
Actually since Get does not use any cmdlets from the DNS module we can just remove `Assert-Module` 

https://github.com/dsccommunity/xDnsServer/blob/2a8ea7f52c1a29637a5fc895105b8defeddbe4d8/source/DSCResources/MSFT_xDnsServerSetting/MSFT_xDnsServerSetting.psm1#L22

and

https://github.com/dsccommunity/xDnsServer/blob/2a8ea7f52c1a29637a5fc895105b8defeddbe4d8/source/DSCResources/MSFT_xDnsServerSetting/MSFT_xDnsServerSetting.psm1#L225

**Comment by rchristman89** [comment]
Most of the resources in XDnsServer do require it. So I was thinking it needed to be added to those. 

This was specifically opened to not error out in Get, is that something we can support?

**Comment by johlju** [comment]
Yes to both. For all the resource; Get-TargetResource should be able to return the result without throwing an error. It should always return the correct hashtable (all the properties in the schema), either with the actual current values or with null values for those properties it cannot fetch. Key properties should always be set to the configuration values and if there is an Ensure property it should be set to Absent if the Key values does not exist in the current state.

**Comment by rchristman89** [comment]
Awesome. I will look at doing that.

**Comment by ahpooch** [comment]
# Clarifying the Current State of the Issue
I’d like to bring clarity to this problem in the current context. As I understand it, the original request was to enable the use of DnsServerDsc resources in environments where the DnsServer module is not installed when running DSC configurations in Test mode. The initial question specifically concerned xDNSServerForwarder, which is now the MOF-Based resource DSC_DnsServerForwarder.

@johlju suggested modifying Get-TargetResource, which is only used in MOF-Based resources.

Class-Based resources, on the other hand, implement a `hidden [Microsoft.Management.Infrastructure.CimInstance] GetResourceRecord()` method.
The following code in this method would, as I understand it, fail on systems where the DnsServer module is missing:
```powershell
$record = Get-DnsServerResourceRecord @dnsParameters -ErrorAction SilentlyContinue | Where-Object {  
    $_.RecordData.IPv4Address -eq $this.IPv4Address  
}  
```
The `Test()` method relies on `GetResourceRecord()`, and this is precisely the method intended to work even when the DnsServer module is absent.

## What could be done
### For Class-Based Resources:
Modify GetResourceRecord() to:
- Check for the availability of Get-DnsServerResourceRecord.
- Return an empty result if the function is unavailable (e.g., module not installed).

### For MOF-Based Resources:
Enhance Assert-Module (from DscResource.Common) by:
- Adding a parameter like -ErrorAction or -ForceError to:
  - Suppress errors in Get-TargetResource.
  - Enforce errors in Set-TargetResource (where the module is mandatory).

## Additional Observations:
Not all MOF-Based resources implement this Assert-Module check, which ideally should be corrected by adding it where missing.

### Assert-Module function usage in MOF-Based Resources
Count of resources that use Assert-Module -ModuleName 'DNSServer': 8
Count of resources unit tests that use Assert-Module (Mock): 7
Count of resources that not use Assert-Module -ModuleName 'DNSServer': 5
Count of resources unit tests that not use Assert-Module (Mock): 6

### List of files that utilize Get-TargetResource and optionaly Assert-Module
!Special CASE!: Resource using Assert-Module but Assert-Module Absent in its Tests.
- DSC_DnsServerDiagnostics.psm1
  YES -ModuleName 'DNSServer'
- DSC_DnsServerDiagnostics.Tests.ps1
  NO

- DSC_DnsServerADZone.psm1
  YES -ModuleName 'DNSServer'
- DSC_DnsServerADZone.Tests.ps1
  YES (Mock)

- DSC_DnsServerPrimaryZone.psm1
  YES -ModuleName 'DNSServer'
- DSC_DnsServerPrimaryZone.Tests.ps1
  YES (Mock)

- DSC_DnsServerZoneTransfer.psm1
  YES -ModuleName 'DNSServer'
- DSC_DnsServerZoneTransfer.Tests.ps1
  YES (Mock)

- DSC_DnsServerRootHint.psm1
  YES -ModuleName 'DNSServer'
- DSC_DnsServerRootHint.Tests.ps1
  YES (Mock)

- DSC_DnsServerSecondaryZone.psm1
  YES -ModuleName 'DNSServer'
- DSC_DnsServerSecondaryZone.Tests.ps1
  YES (Mock)

- DSC_DnsServerSetting.psm1
  YES -ModuleName 'DNSServer'
- DSC_DnsServerSetting.Tests.ps1
  YES (Mock)

- DSC_DnsServerSettingLegacy.psm1
  YES -ModuleName 'DNSServer'
- DSC_DnsServerSettingLegacy.Tests.ps1
  YES (Mock)

- DSC_DnsServerClientSubnet.psm1
  NO
- DSC_DnsServerClientSubnet.Tests.ps1
  NO

- DSC_DnsServerConditionalForwarder.psm1
  NO
- DSC_DnsServerConditionalForwarder.Tests.ps1
  NO

- DSC_DnsServerForwarder.psm1
  NO
- DSC_DnsServerForwarder.Tests.ps1
  NO

- DSC_DnsServerZoneAging.psm1
  NO
- DSC_DnsServerZoneAging.Tests.ps1
  NO

- DSC_DnsServerZoneScope.psm1
  NO
- DSC_DnsServerZoneScope.Tests.ps1
  NO
